### PR TITLE
Use new isDeleted property for deleted actions

### PR DESCRIPTION
**Pull request by arosiclair**
- HELD on https://github.com/Expensify/Web-Expensify/pull/41106

### Details
We are now marking report actions as deleted by setting the `deleted` timestamp in their `message` array. We'll hide and ignore these actions like we did with deleted comments. For legacy support, we're still hiding `ADDCOMMENT` actions that have an empty message.

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/367794

### Tests

**Deleting a money request**

1. Sign into two accounts user A and B
2. As user A, start a new money request from user B such that there is only one expense on the report
3. Verify there's a report preview in their DM
4. Navigate to Report > Expense
5. Select Delete request in the 3 dot menu
6. Verify you're navigated back to the DM and the report preview is deleted

**Changing a workspace money request into an invoice**

1. Sign in as a collect policy admin
2. Open the policy settings page and run this JS snippet to enable policy expense chats

```javascript
p = Policy.getCurrent();
p.policy.isPolicyExpenseChatEnabled = "true";
p.save();
```

4. Invite a new member
5. As the member, request money in the workspace chat
6. Sign into OldDot as the member
7. Find the expense report for the money request
8. Go to Details and change the Type to invoice
9. In NewDot, verify the report is removed from the workspace chat

### Offline tests

None

### QA Steps

Same as Tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot_20240305-151619

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot_20240305-111938

</details>

<details>
<summary>iOS: Native</summary>

IMG_ED5F0D911EE3-1

</details>

<details>
<summary>iOS: mWeb Safari</summary>

IMG_0F4C293F2304-1

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-03-05 at 10 51 47 AM

</details>

<details>
<summary>MacOS: Desktop</summary>

Screenshot 2024-03-05 at 11 03 21 AM

</details>

### CODE REVIEW

**Comment by arosiclair** [comment]
@DylanDylann backend changes that are needed for this are on staging so make sure to test against that API. I'm not sure if you can test the second part (Changing a workspace money request into an invoice), but if you can't, that's okay. The internal engineer can take a look instead.

**Comment by DylanDylann** [comment]
Let's me try

**Comment by DylanDylann** [comment]
@arosiclair Do we only use isDeleted `property` for money request action?

**Comment by arosiclair** [comment]
> @arosiclair Do we only use `deleted` property for money request action?

With the new backend changes, yes.

**Comment by DylanDylann** [comment]
@arosiclair Should we update `deleted` property of money report action in the IOU report? Currently, BE also doesn't support that

**Comment by DylanDylann** [comment]
In case, we have many money requests. If we delete a request money, we need to hide it in the IOU report instead of hiding the report preview in the main report (1:1 chat)

**Comment by arosiclair** [comment]
> @arosiclair Should we update deleted property of money report action in the IOU report? Currently, BE also doesn't support that

That shouldn't be necessary atm.

> In case, we have many money requests. If we delete a request money, we need to hide it in the IOU report instead of hiding the report preview in the main report (1:1 chat)

Ah yes to clarify, you must test with a new money request that only has one expense. Deleting the last expense will also delete the report and the report preview in the chat. I just updated the tests to reflect that. 

The scenario of deleting one expense with others still left on the report should be unaffected.

**Comment by DylanDylann** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by DylanDylann — APPROVED** [review]

File: src/libs/ReportActionsUtils.ts [hunk 1]
```diff
@@ -95,9 +95,12 @@ function isCreatedAction(reportAction: OnyxEntry<ReportAction>): boolean {
 }
 
 function isDeletedAction(reportAction: OnyxEntry<ReportAction | OptimisticIOUReportAction>): boolean {
-    // A deleted comment has either an empty array or an object with html field with empty string as value
     const message = reportAction?.message ?? [];
-    return message.length === 0 || message[0]?.html === '';
+
+    // A legacy deleted comment has either an empty array or an object with html field with empty string as value
+    const isLegacyDeletedComment = reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT && (!message.length || !message[0]?.html);
```
Comment (grgia) [inline src/libs/ReportActionsUtils.ts:101]: Double checking, there's no regressions we might cause by adding`reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT`? Is it only comments that we have delete support for?

File: src/libs/actions/IOU.ts [hunk 2]
```diff
@@ -3087,12 +3087,10 @@ function deleteMoneyRequest(transactionID: string, reportAction: OnyxTypes.Repor
             onyxMethod: Onyx.METHOD.MERGE,
             key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${chatReport?.reportID}`,
             value: {
-                [reportPreviewAction?.reportActionID ?? '']: shouldDeleteIOUReport
```
Comment (grgia) [inline src/libs/actions/IOU.ts:3090]: why are we removing this `shouldDeleteIOUReport` check here?

**Review by grgia — DISMISSED** [review]
LGTM. Will wait on comments to merge, but none are blockers

File: src/libs/ReportActionsUtils.ts (hunk 1 above)
Comment (grgia) [reply src/libs/ReportActionsUtils.ts:101]: for example, money requests?

Comment (DylanDylann) [reply src/libs/ReportActionsUtils.ts:101]: @grgia Thanks for your reminder. As I understand, @arosiclair added this condition to filter request money report action from `isLegacyDeletedComment` check. But It seems we missed [this case](https://github.com/Expensify/App/pull/36747#issuecomment-1994941714). We only should ignore `isLegacyDeletedComment` check if It is the only request in report

File: src/libs/actions/IOU.ts (hunk 2 above)
Comment (DylanDylann) [reply src/libs/actions/IOU.ts:3090]: @grgia Currently, if `shouldDeleteIOUReport` we will delete PREVIEW report action in the main chat (1:1 chat). With this update, we will not remove PREVIEW report action from the report action list, we only add `deleted` field as a flag to mark this request is deleted

File: src/libs/ReportActionsUtils.ts (hunk 1 above)
Comment (DylanDylann) [reply src/libs/ReportActionsUtils.ts:101]: @arosiclair Should we update this?

Comment (arosiclair) [reply src/libs/ReportActionsUtils.ts:101]: > Double checking, there's no regressions we might cause by addingreportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT? Is it only comments that we have delete support for?

Only report comments use this pattern of clearing the `message` so it shouldn't cause any issues. There are many actions with empty messages by default ([see here](https://expensify.slack.com/archives/C03TQ48KC/p1708114308456789?thread_ts=1708030880.121819&cid=C03TQ48KC)), so just checking for an empty message on any type of action was not correct.

> @arosiclair Should we update this?

I'm not sure understand your concern. When you're testing deleting expenses on a report, are you noticing any bugs or regressions? If so, can you write repro steps with your expected and actual outcomes?

File: src/libs/actions/IOU.ts (hunk 2 above)
Comment (arosiclair) [reply src/libs/actions/IOU.ts:3090]: Yeah this is just a change to stop deleting the report preview action in the `successData` for `DeleteMoneyRequest`

File: src/libs/ReportActionsUtils.ts (hunk 1 above)
Comment (DylanDylann) [reply src/libs/ReportActionsUtils.ts:101]: @arosiclair Currently, I don't see any regression. But let's see the video below






https://github.com/Expensify/App/assets/141406735/5521a941-281d-4ab2-aef7-b8e9998e5e51



I have 2 money requests. After I delete the second request, isDeletedAction still returns false

Comment (arosiclair) [reply src/libs/ReportActionsUtils.ts:101]: Okay I see what you mean thanks for clarifying. It does look like a bug for those actions to still display after being deleted. We'll need more backend changes to address deleted expense actions, but that can be done later. 

I'll remove the ADDCOMMENT check so these get hidden and re-add it later once that's fixed.

**Review by DylanDylann — APPROVED** [review]
@grgia All yours

**Comment by DylanDylann** [comment]
@arosiclair Test fails. Please help to merge main

**Comment by DylanDylann** [comment]
_I shouldn't approve again_

@Julesssss Please ignore this PR. @grgia is assigned before

**Review by grgia — DISMISSED** [review]

**Comment by grgia** [comment]
@arosiclair bump on merge main :)

**Comment by arosiclair** [comment]
We actually need a strict equality check for the message html (`message[0].html === ''`) because `CREATED` actions have `html` undefined in their message array for some reason. That doesn't seem right but that's a separate issue. Ready for another review.

**Review by grgia — APPROVED** [review]

**Comment by grgia** [comment]
@DylanDylann can you confirm if you've retested

**Comment by DylanDylann** [comment]
@grgia Yeah, everything work well

**Pull request merged**